sdk/core/Azure.Core/src/Shared/OperationInternal.cs

```diff
@@ -0,0 +1,236 @@
+                }
+            }
+            set
+            {
+                if (value is null)
+                {
+                    throw new ArgumentNullException(nameof(Value));
+                }
+
+                _value = value;
+                HasValue = true;
+            }
+        }
+
+        public Response RawResponse { get; set; }
+
+        public string OperationTypeName { get; set; }
+
+        public TimeSpan DefaultPollingInterval { get; set; }
+
+        protected ClientDiagnostics ClientDiagnostics { get; }
+
+        protected RequestFailedException OperationFailedException { get; set; }
+
+        public async ValueTask<Response> UpdateStatusAsync(CancellationToken cancellationToken) =>
+            await UpdateStatusAsync(async: true, cancellationToken).ConfigureAwait(false);
+
+        public Response UpdateStatus(CancellationToken cancellationToken) =>
+            UpdateStatusAsync(async: false, cancellationToken).EnsureCompleted();
+
+        public async ValueTask<Response<TResult>> WaitForCompletionAsync(CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                Response response = await UpdateStatusAsync(cancellationToken).ConfigureAwait(false);
+
+                if (HasCompleted)
+                {
+                    return Response.FromValue(Value, response);
+                }
+
+                var pollingInterval = response.Headers.TryGetValue(RetryAfterHeaderName, out string retryAfterValue)
+                    && int.TryParse(retryAfterValue, out int delayInSeconds)
+                    ? TimeSpan.FromSeconds(delayInSeconds) : DefaultPollingInterval;
+
+                await Task.Delay(pollingInterval, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        public async ValueTask<Response<TResult>> WaitForCompletionAsync(TimeSpan pollingInterval, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                Response response = await UpdateStatusAsync(cancellationToken).ConfigureAwait(false);
+
+                if (HasCompleted)
+                {
+                    return Response.FromValue(Value, response);
+                }
+
+                await Task.Delay(pollingInterval, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        protected virtual async ValueTask<Response> UpdateStateAsync(bool async, CancellationToken cancellationToken)
+        {
+            Response response = async
+                ? await _operation.GetResponseAsync(cancellationToken).ConfigureAwait(false)
+                : _operation.GetResponse(cancellationToken);
+
+            var state = _operation.UpdateState(response);
+
+            SetState(state, response);
+
+            return response;
+        }
+
+        protected void SetState(OperationState<TResult> state, Response response)
+        {
+            lock (SetStateLock)
```

Comment:
[Description about the race condition](https://github.com/Azure/azure-sdk-for-net/pull/18354#pullrequestreview-581591037).

I'm wondering if we truly need this lock. Actually, it won't prevent the race condition from happening:
- Thread A and B call `UpdateStatus` almost at the same time.
- Thread B gets a response from the service saying that operation completed successfully.
- Thread A gets a response from the service saying that operation is still running, but we got it later.
- B ends up entering the lock first. Sets `RawResponse` and `Value`.
- A enters the lock later, only updating `RawResponse`.
- Now we have the `Value` we got from the completed operation, but `RawResponse` is not the latest one.

We're allowing users to make a service call with `UpdateStatus` even after the operation has completed, so the latest raw response can be obtained with another method call in case users need it.

I can think of two ways to prevent this race condition from happening:
- We could put the service call inside a lock, forcing one call to happen at a time. This seems like a big bottleneck to me.
- We could keep the lock and ignore raw responses with "pending status" that we got after the operation has completed. I wonder if this is the right thing to do, since `RawResponse` is supposed to be "the last response received from the service".

Reply:
>We could put the service call inside a lock, forcing one call to happen at a time. This seems like a big bottleneck to me.

My first reaction was to go with the second approach and it might still be the right one short-term. But then I started thinking. What would be the benefit of having multiple UpdateStatus calls hitting the service at the same time?

Reply:
That's a good point. However, we're not short-circuiting `UpdateStatus` calls anymore, so we would need to make all these calls to the service eventually. Keeping them in line in a lock feels a bit slow.

We stopped short-circuiting `UpdateStatus` because we realized `RawResponse` can be updated by other methods as well (such as `CancelOperation`). If we skip the service call, we'll return the wrong response and make it impossible for users to get the latest status.

I think we can go with the first approach and start short-circuiting `UpdateStatus` again if we did something like:
```cs
public Response RawResponse { get; set; }

private Response UpdateStatusRawResponse { get; set; }

Response UpdateStatus(...)
{
  lock (...)
  {
    if (HasCompleted)
    {
      RawResponse = UpdateStatusRawResponse;
      return UpdateStatusRawResponse;
    }
    ... // make service call, update state, set UpdateStatusRawResponse
  }
}
```

Reply:
Let's not short-circuit for now and not force a single request at a time as well. Start simple.

Reply:
You think we should go with the second approach then? Or just ignore the race condition for now?

Reply:
Yeah, let's ignore it for now. As long as the UpdateStatus calls return the same thing we are good. The GetRawResponse() is defined as the "last response" and in the parallel case it's non-deterministic by-definition.